Refine sandboxed user perms for query builder access

escherize opened the pull request:
This is a fix for the scenario where if a user has a sandboxed adhoc query, they can use it to read from tables they are blocked on. 😱

If you use the query builder to setup a sandboxed query that joins to a table you are `can-view = blocked` on, we no longer grant you `create-queries` access to it when doing the sandboxing permissions calculations.

How:

- Limit create-queries permissions to unblocked tables only
- Check user permissions for each table before granting query builder access
- Prevent querying of blocked joined tables from query builder for sandboxed users

escherize committed:
Refine sandboxed user perms for query builder access

- Limit create-queries permissions to unblocked tables only
- Check user permissions for each table before granting query builder access
- Prevent querying of blocked joined tables from query builder for sandboxed users

johnswanson commented on enterprise/backend/src/metabase_enterprise/sandbox/query_processor/middleware/row_level_restrictions.clj:
```diff
@@ -259,9 +260,24 @@
+                                                     (get table-id->db-id table-id)
+                                                     table-id))
+                                      table-ids)
+          create-queries-perms (zipmap unblocked-table-ids (repeat :query-builder))]
```
Is there a reason to grant `:query-builder` to only the unblocked tables, but `:unrestricted` to all of them? I think it might be better to grant `:perms/view-data :unrestricted` to only the unblocked tables, and `:query-builder` to all of them.

Say you already have `:query-builder` access to all tables but are `:blocked` on all but one, where you have sandbox access. We don't want to override that block, right?

escherize replied:
I think you're right. I will switch that around and include it in the upcoming tests

escherize committed:
Adjust permissions for sandboxed users

- Grant view-data permissions only for unblocked tables
- Revert create-queries permissions to all tables in sandbox
- Remove unnecessary intermediate variable

escherize committed:
when sandboxing we no longer grant unrestricted view perms for blocked tables
